Aplib.Core/ExposedQueue.cs

```diff
@@ -0,0 +1,193 @@
+        /// <param name="arrayIndex">The start index of the range to copy.</param>
+        /// <param name="endIndex">The end index of the range to copy.</param>
+        /// <returns>The circular array as a normal array</returns>
+        public void CopyTo(T[] array, int arrayIndex, int endIndex)
+        {
+            if (arrayIndex < 0 || arrayIndex >= Count)
+                throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+            if (endIndex < 0 || endIndex >= Count)
+                throw new ArgumentOutOfRangeException(nameof(endIndex));
+            if (arrayIndex > endIndex)
+                throw new ArgumentException("Start index must be less than or equal to end index.");
+            if (array.Length < endIndex - arrayIndex + 1)
+                throw new ArgumentException("Array is too small to copy the range.");
+
+            for (int i = 0; i < endIndex - arrayIndex + 1; i++)
+                array[i] = this[arrayIndex + i];
+        }
+
+        /// <inheritdoc/>
+        public void CopyTo(T[] array, int arrayIndex) => CopyTo(array, arrayIndex, arrayIndex + Count - 1);
+
+        /// <summary>
+        /// Converts the ExposedQueue to an array.
+        /// </summary>
+        /// <param name="start">The start index of the range to convert.</param>
+        /// <param name="end">The end index of the range to convert.</param>
+        /// <returns>An array containing the elements within the specified range.</returns>
+        public T[] ToArray(int start, int end)
+        {
+            Console.WriteLine($"Start: {start}, End: {end}");
+            if (start < 0 || start >= Count)
+                throw new ArgumentOutOfRangeException(nameof(start), "Start index must be within the bounds of the array.");
+            if (end < 0 || end >= Count)
+                throw new ArgumentOutOfRangeException(nameof(end), "End index must be within the bounds of the array.");
+            T[] result = new T[end - start + 1];
+            CopyTo(result, start, end);
+            return result;
+        }
+
+        /// <summary>
+        /// Converts the ExposedQueue to an array. Only returns the used slots.
+        /// </summary>
+        /// <returns>An array containing the elements within the specified range.</returns>
+        public T[] ToArray() => ToArray(0, Count - 1);
+
+        /// <inheritdoc/>
+        public void Clear()
+        {
+            for (int i = 0; i < MaxCount; i++)
+                _array[i] = default!;
+            _head = MaxCount - 1;
+            Count = 0;
+        }
+
+        /// <inheritdoc/>
+        public bool Contains(T item)
+        {
+            for (int i = 0; i < Count; i++)
+                if (this[i]!.Equals(item))
+                    return true;
+            return false;
+        }
+
+        /// <inheritdoc/>
+        public bool Remove(T item)
+        {
+            for (int i = 0; i < Count; i++)
+                if (this[i]!.Equals(item))
```

Comment:
Is the null-forgiving operator necessary here?
```suggestion
                if (this[i].Equals(item))
```

Reply:
It is, nvm